# Домашнє завдання 2 (Тема 3). Завантаження даних, основи SQL, DQL та EDA

**Датасет:** варіант **A — NYC TLC Yellow Taxi Trip Records**, січень 2024
(`yellow_tripdata_2024-01.parquet`, ≈2.96 млн поїздок), контрольована випадкова вибірка **100 000** рядків (`random_state=42`).

**Чому саме цей датасет:** він єдиний із чотирьох, що природно покриває всі вимоги до DDL —
грошові суми (`NUMERIC(10,2)`), момент події з часовою зоною (`TIMESTAMPTZ`), календарну дату (`DATE`),
справжню двійкову ознаку (`store_and_fwd_flag` Y/N → `BOOLEAN`), а також має реальні пропуски, спеціальні коди
(`RatecodeID = 99`), від'ємні суми та інші аномалії для аудиту.

**Принцип роботи:** pandas використовується *лише* для перетворення Parquet → CSV. Усе завантаження (COPY FROM STDIN),
очищення, аудит якості та EDA виконуються SQL-запитами в PostgreSQL (pgserver).

Структура: 1) Setup і завантаження → 2) Staging і DDL → 3) Data-quality audit → 4) Розподіли та аномалії →
5) Десять DQL/EDA-запитів → 6) Reflection.

## Завдання 1. Setup і завантаження

### 1.1. Встановлення залежностей, запуск PostgreSQL, перевірка підключення

Основний шлях — `pgserver`, як у завданні. Але в `pgserver` немає збірок під актуальну версію Python у Google Colab
(`pip` видає *No matching distribution found for pgserver*). Тому клітинка спершу пробує `pgserver`, а якщо його
неможливо встановити — піднімає системний PostgreSQL через `apt-get` (так само, як у ДЗ 02).
Решта notebook не залежить від того, який варіант спрацював: далі використовується лише `engine`.

In [1]:
%pip install -q psycopg2-binary sqlalchemy pandas pyarrow
%pip install -q pgserver

import subprocess

import pandas as pd
from sqlalchemy import create_engine, text

try:
    import pgserver

    pg = pgserver.get_server('/tmp/hw3_pg', cleanup_mode='stop')
    db_uri = pg.get_uri()
    print('PostgreSQL backend: pgserver')
except ImportError:
    # Fallback для Colab: системний PostgreSQL з apt (як у ДЗ 02)
    for cmd in [
        'apt-get update -qq',
        'apt-get install -y -qq postgresql postgresql-contrib > /dev/null',
        'service postgresql start',
        "sudo -u postgres psql -q -c \"ALTER USER postgres WITH PASSWORD 'postgres';\"",
    ]:
        subprocess.run(cmd, shell=True, check=True)
    db_uri = 'postgresql://postgres:postgres@127.0.0.1:5432/postgres'
    print('PostgreSQL backend: apt-get postgresql')

# SQLAlchemy 2.1+ за замовчуванням обирає драйвер psycopg (v3) — явно вказуємо встановлений psycopg2
engine = create_engine(db_uri.replace('postgresql://', 'postgresql+psycopg2://', 1), future=True)

with engine.connect() as conn:
    version = conn.execute(text('SELECT version()')).scalar_one()
    print(version)

Note: you may need to restart the kernel to use updated packages.


Note: you may need to restart the kernel to use updated packages.


PostgreSQL backend: pgserver
PostgreSQL 16.2 on aarch64-apple-darwin23.5.0, compiled by Apple clang version 15.0.0 (clang-1500.0.40.1), 64-bit


Допоміжні функції: `q()` виконує DQL-запит і повертає DataFrame, `run()` виконує DDL/DML у транзакції.
Сесійна часова зона PostgreSQL фіксується як `America/New_York`, щоб `TIMESTAMPTZ` відображались у місцевому часі поїздок
(зберігаються вони все одно як абсолютний момент в UTC).

In [2]:
from pathlib import Path

pd.set_option('display.max_columns', 50)
pd.set_option('display.width', 200)

with engine.begin() as conn:
    conn.execute(text("ALTER ROLE CURRENT_USER SET timezone = 'America/New_York'"))
engine.dispose()  # нові з'єднання підхоплять налаштування timezone


def q(sql: str) -> pd.DataFrame:
    """Виконати SELECT і повернути результат як DataFrame.

    pandas приводить TIMESTAMPTZ до UTC, тому для читабельності показуємо їх у місцевому часі NY.
    """
    df = pd.read_sql(text(sql), engine)
    for col in df.select_dtypes(include=['datetimetz']).columns:
        df[col] = df[col].dt.tz_convert('America/New_York')
    return df


def run(sql: str) -> None:
    """Виконати DDL/DML (один або кілька операторів) в одній транзакції."""
    with engine.begin() as conn:
        conn.execute(text(sql))


q("SELECT current_setting('TimeZone') AS session_tz, current_database() AS db")

,session_tz,db
0,America/New_York,postgres


### 1.2. Отримання даних з офіційного джерела

Джерело: сторінка [TLC Trip Record Data](https://www.nyc.gov/site/tlc/about/tlc-trip-record-data.page) →
*2024 → January → Yellow Taxi Trip Records (PARQUET)*. Посилання на файл веде на офіційний CDN TLC (CloudFront).
Додатково завантажуємо офіційний довідник зон `taxi_zone_lookup.csv` з тієї ж сторінки (для розшифровки `LocationID`).

Замість `head(100_000)` береться **випадкова вибірка з фіксованим seed**: файл упорядкований приблизно за часом,
тому `head()` дав би лише перші 1–2 дні місяця і спотворив би розподіли за днями тижня та годинами.
Беремо всі 19 колонок файлу — це потрібно, щоб звірити `total_amount` із сумою його складових.

In [3]:
from urllib.request import urlretrieve

DATA_URL  = 'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2024-01.parquet'
ZONES_URL = 'https://d37ci6vzurychx.cloudfront.net/misc/taxi_zone_lookup.csv'

# У Colab працюємо в /content, локально — у ./data
DATA_DIR = Path('/content') if Path('/content').exists() else Path.cwd() / 'data'
DATA_DIR.mkdir(parents=True, exist_ok=True)

parquet_path = DATA_DIR / 'hw3_taxi_2024_01.parquet'
csv_path     = DATA_DIR / 'hw3_taxi_sample.csv'
zones_path   = DATA_DIR / 'taxi_zone_lookup.csv'

if not parquet_path.exists():
    urlretrieve(DATA_URL, parquet_path)
if not zones_path.exists():
    urlretrieve(ZONES_URL, zones_path)

columns = [
    'VendorID', 'tpep_pickup_datetime', 'tpep_dropoff_datetime',
    'passenger_count', 'trip_distance', 'RatecodeID', 'store_and_fwd_flag',
    'PULocationID', 'DOLocationID', 'payment_type',
    'fare_amount', 'extra', 'mta_tax', 'tip_amount', 'tolls_amount',
    'improvement_surcharge', 'total_amount', 'congestion_surcharge', 'Airport_fee',
]

df_full = pd.read_parquet(parquet_path, columns=columns)
df = df_full.sample(n=100_000, random_state=42).sort_values('tpep_pickup_datetime')
df.to_csv(csv_path, index=False)

print('full month:', df_full.shape, '| sample:', df.shape)
del df_full

full month: (2964624, 19) | sample: (100000, 19)


Перші рядки контрольованого CSV — саме в такому текстовому вигляді дані потраплять у staging (зверніть увагу на `2.0` у `passenger_count` і порожні поля):

In [4]:
with open(csv_path, encoding='utf-8') as f:
    for _ in range(4):
        print(f.readline().rstrip())

VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee
2,2023-12-31 23:56:46,2024-01-01 00:12:06,2.0,2.38,1.0,N,236,142,1,15.6,1.0,0.5,1.0,0.0,1.0,21.6,2.5,0.0
2,2023-12-31 23:58:35,2024-01-01 00:13:06,6.0,8.39,1.0,N,138,217,2,33.1,6.0,0.5,0.0,0.0,1.0,42.35,0.0,1.75
2,2024-01-01 00:00:17,2024-01-01 00:07:46,1.0,2.35,1.0,N,161,75,1,12.1,1.0,0.5,3.42,0.0,1.0,20.52,2.5,0.0


### 1.3. Staging-таблиця та COPY FROM STDIN

Усі сирі колонки мають тип `TEXT`: так імпорт не впаде на `2.0` у цілочисельному полі, порожньому значенні чи
некоректній даті — типізація і валідація відбуваються пізніше, у SQL.
Важливий нюанс `COPY ... FORMAT csv`: **неекрановане порожнє поле стає SQL `NULL`**, а `""` в лапках — порожнім рядком `''`.
Тому пропуски з pandas (`NaN` → порожнє поле) у staging одразу виглядають як `NULL`, а `NULLIF(TRIM(x), '')` під час
очищення додатково страхує від порожніх рядків/пробілів.

In [5]:
run("""
DROP TABLE IF EXISTS hw3_taxi_trips CASCADE;
DROP TABLE IF EXISTS hw3_taxi_staging CASCADE;

CREATE TABLE hw3_taxi_staging (
    vendor_id_raw             TEXT,
    pickup_raw                TEXT,
    dropoff_raw               TEXT,
    passenger_count_raw       TEXT,
    trip_distance_raw         TEXT,
    rate_code_id_raw          TEXT,
    store_and_fwd_flag_raw    TEXT,
    pu_location_id_raw        TEXT,
    do_location_id_raw        TEXT,
    payment_type_raw          TEXT,
    fare_amount_raw           TEXT,
    extra_raw                 TEXT,
    mta_tax_raw               TEXT,
    tip_amount_raw            TEXT,
    tolls_amount_raw          TEXT,
    improvement_surcharge_raw TEXT,
    total_amount_raw          TEXT,
    congestion_surcharge_raw  TEXT,
    airport_fee_raw           TEXT
);
""")

In [6]:
STAGING_COLUMNS = """
    vendor_id_raw, pickup_raw, dropoff_raw, passenger_count_raw, trip_distance_raw,
    rate_code_id_raw, store_and_fwd_flag_raw, pu_location_id_raw, do_location_id_raw,
    payment_type_raw, fare_amount_raw, extra_raw, mta_tax_raw, tip_amount_raw,
    tolls_amount_raw, improvement_surcharge_raw, total_amount_raw,
    congestion_surcharge_raw, airport_fee_raw
"""

raw_conn = engine.raw_connection()
try:
    with raw_conn.cursor() as cur:
        with open(csv_path, 'r', encoding='utf-8', newline='') as file:
            cur.copy_expert(
                f"""
                COPY hw3_taxi_staging ({STAGING_COLUMNS})
                FROM STDIN
                WITH (FORMAT csv, HEADER true, DELIMITER ',')
                """,
                file
            )
    raw_conn.commit()
finally:
    raw_conn.close()

q('SELECT COUNT(*) AS loaded_rows FROM hw3_taxi_staging')

,loaded_rows
0,100000


Довідник зон — маленький (265 рядків), чистий офіційний файл із гарантованою структурою, тому його завантажуємо
тим самим `COPY FROM STDIN` одразу в типізовану таблицю. Зверніть увагу: `Unknown` / `N/A` у довіднику — це
спеціальні маркери невідомої локації (ID 264, 265), а не справжні назви районів.

In [7]:
run("""
DROP TABLE IF EXISTS hw3_taxi_zones CASCADE;
CREATE TABLE hw3_taxi_zones (
    location_id  SMALLINT PRIMARY KEY,
    borough      TEXT NOT NULL,
    zone         TEXT NOT NULL,
    service_zone TEXT
);
""")

raw_conn = engine.raw_connection()
try:
    with raw_conn.cursor() as cur, open(zones_path, 'r', encoding='utf-8', newline='') as file:
        cur.copy_expert(
            "COPY hw3_taxi_zones FROM STDIN WITH (FORMAT csv, HEADER true, DELIMITER ',')",
            file
        )
    raw_conn.commit()
finally:
    raw_conn.close()

q('SELECT COUNT(*) AS zones FROM hw3_taxi_zones')

,zones
0,265


## Завдання 2. Staging і DDL

### 2.1. Семантика колонок, джерельний формат, тип і правила очищення

Коди — за офіційним [Data Dictionary Yellow Taxi](https://www.nyc.gov/assets/tlc/downloads/pdf/data_dictionary_trip_records_yellow.pdf).

| Колонка (clean) | Семантика | Джерельний формат (CSV) | Тип | Правила очищення / обмеження |
|---|---|---|---|---|
| `trip_id` | технічний surrogate-ключ | — | `BIGINT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY` | генерується БД; природного унікального ID поїздки у файлі немає |
| `vendor_id` | провайдер TPEP-системи (1 = Creative Mobile, 2 = Curb/VeriFone, 6 = Myle, 7 = Helix) | ціле `2` | `SMALLINT NOT NULL` | `NULLIF(TRIM(..),'')::SMALLINT`; `CHECK (vendor_id IN (1,2,6,7))` — закритий довідник кодів |
| `pickup_at`, `dropoff_at` | моменти посадки/висадки | локальний час NY **без** зони: `2024-01-20 13:31:30` | `TIMESTAMPTZ NOT NULL` | `::TIMESTAMP AT TIME ZONE 'America/New_York'` — явна локалізація, після якої значення стає абсолютним моментом; без цих полів поїздка не має змісту → `NOT NULL`; `CHECK (dropoff_at >= pickup_at)` |
| `pickup_date` | календарна дата посадки (місцева) | — (похідне) | `DATE GENERATED ALWAYS AS (...) STORED` | `(pickup_at AT TIME ZONE 'America/New_York')::DATE` — дата без часу для групувань по днях; обчислюється БД, тому не може розійтися з `pickup_at` |
| `passenger_count` | кількість пасажирів (вводить водій) | float з pandas: `2.0`, порожньо = пропуск | `SMALLINT` (nullable) | `::NUMERIC::SMALLINT` (бо `'2.0'::SMALLINT` падає); `NULL` залишається `NULL`; `0` зберігаємо (це реальне, хоч і сумнівне, значення водія); `CHECK (passenger_count BETWEEN 0 AND 9)` |
| `trip_distance` | відстань за таксометром, милі | `17.14` | `NUMERIC(10,2) NOT NULL` | точність до сотих милі як у джерелі; `CHECK (trip_distance >= 0)` |
| `rate_code_id` | тариф (1 Standard, 2 JFK, 3 Newark, 4 Nassau/Westchester, 5 Negotiated, 6 Group) | float `1.0`; `99.0` = «Null/unknown» | `SMALLINT` (nullable) | `NULLIF(::NUMERIC::SMALLINT, 99)` — **спецмаркер 99 → SQL NULL**, бо за словником це саме «невідомо»; `CHECK (rate_code_id BETWEEN 1 AND 6)` |
| `store_and_fwd` | чи запис зберігався в пам'яті авто і був переданий пізніше (немає зв'язку) | `Y` / `N` / порожньо | `BOOLEAN` (nullable) | `CASE UPPER(TRIM(..)) WHEN 'Y' THEN TRUE WHEN 'N' THEN FALSE END` — справжня двійкова ознака; невідоме значення → `NULL`, а не `FALSE` |
| `pu_location_id`, `do_location_id` | TLC Taxi Zone посадки/висадки | ціле `132` | `SMALLINT NOT NULL` | `CHECK (... BETWEEN 1 AND 265)` — діапазон офіційного довідника зон (264/265 = unknown/outside NYC) |
| `payment_type` | спосіб оплати (1 Credit card, 2 Cash, 3 No charge, 4 Dispute, 5 Unknown, 6 Voided; 0 — поза словником) | ціле `1` | `SMALLINT NOT NULL` | зберігаємо код як є (включно з `0`, щоб не втратити інформацію); `CHECK (payment_type BETWEEN 0 AND 6)` |
| `fare_amount`, `extra`, `mta_tax`, `tip_amount`, `tolls_amount`, `improvement_surcharge`, `total_amount`, `congestion_surcharge`, `airport_fee` | складові чека та підсумок, USD | float `70.0`, `8.27`; для двох останніх бувають порожні | `NUMERIC(10,2)` | гроші — лише точний десятковий тип до центів (не `FLOAT`); **від'ємні значення не забороняємо** (це коригування/повернення, їх треба бачити в аудиті); `fare_amount`, `total_amount` — `NOT NULL` |

Чого немає у DDL свідомо: `VARCHAR(255)` — текстових полів без бізнес-ліміту у фактах немає, а в довіднику зон використано `TEXT`.

**Правило фільтрації при переносі staging → clean:** відкидаються рядки, де `pickup_at` або `dropoff_at` не парситься
(`NULL`) або `dropoff_at < pickup_at` (фізично неможлива поїздка). Усі інші «підозрілі» рядки (від'ємні суми,
нульова відстань, `passenger_count = 0`) **залишаються** і позначаються на етапі аудиту — видаляти їх лише через
незвичність не можна.

### 2.2. DDL фінальної таблиці

In [8]:
run("""
DROP TABLE IF EXISTS hw3_taxi_trips CASCADE;

CREATE TABLE hw3_taxi_trips (
    trip_id               BIGINT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,

    vendor_id             SMALLINT NOT NULL
        CHECK (vendor_id IN (1, 2, 6, 7)),

    pickup_at             TIMESTAMPTZ NOT NULL,
    dropoff_at            TIMESTAMPTZ NOT NULL,
    pickup_date           DATE GENERATED ALWAYS AS
                              ((pickup_at AT TIME ZONE 'America/New_York')::DATE) STORED,

    passenger_count       SMALLINT
        CHECK (passenger_count BETWEEN 0 AND 9),
    trip_distance         NUMERIC(10, 2) NOT NULL
        CHECK (trip_distance >= 0),
    rate_code_id          SMALLINT
        CHECK (rate_code_id BETWEEN 1 AND 6),
    store_and_fwd         BOOLEAN,

    pu_location_id        SMALLINT NOT NULL
        CHECK (pu_location_id BETWEEN 1 AND 265),
    do_location_id        SMALLINT NOT NULL
        CHECK (do_location_id BETWEEN 1 AND 265),

    payment_type          SMALLINT NOT NULL
        CHECK (payment_type BETWEEN 0 AND 6),

    fare_amount           NUMERIC(10, 2) NOT NULL,
    extra                 NUMERIC(10, 2),
    mta_tax               NUMERIC(10, 2),
    tip_amount            NUMERIC(10, 2),
    tolls_amount          NUMERIC(10, 2),
    improvement_surcharge NUMERIC(10, 2),
    total_amount          NUMERIC(10, 2) NOT NULL,
    congestion_surcharge  NUMERIC(10, 2),
    airport_fee           NUMERIC(10, 2),

    CHECK (dropoff_at >= pickup_at)
);
""")

q("""
SELECT column_name, data_type, is_nullable, is_generated
FROM information_schema.columns
WHERE table_name = 'hw3_taxi_trips'
ORDER BY ordinal_position
""")

,column_name,data_type,is_nullable,is_generated
0,trip_id,bigint,NO,NEVER
1,vendor_id,smallint,NO,NEVER
2,pickup_at,timestamp with time zone,NO,NEVER
3,dropoff_at,timestamp with time zone,NO,NEVER
4,pickup_date,date,YES,ALWAYS
5,passenger_count,smallint,YES,NEVER
6,trip_distance,numeric,NO,NEVER
7,rate_code_id,smallint,YES,NEVER
8,store_and_fwd,boolean,YES,NEVER
9,pu_location_id,smallint,NO,NEVER


### 2.3. Очищення: `INSERT INTO ... SELECT` зі staging

In [9]:
run("""
WITH parsed AS (
    SELECT
        NULLIF(TRIM(vendor_id_raw), '')::SMALLINT                       AS vendor_id,
        NULLIF(TRIM(pickup_raw),  '')::TIMESTAMP
            AT TIME ZONE 'America/New_York'                              AS pickup_at,
        NULLIF(TRIM(dropoff_raw), '')::TIMESTAMP
            AT TIME ZONE 'America/New_York'                              AS dropoff_at,
        NULLIF(TRIM(passenger_count_raw), '')::NUMERIC::SMALLINT         AS passenger_count,
        NULLIF(TRIM(trip_distance_raw), '')::NUMERIC(10, 2)              AS trip_distance,
        NULLIF(NULLIF(TRIM(rate_code_id_raw), '')::NUMERIC::SMALLINT, 99) AS rate_code_id,
        CASE UPPER(TRIM(store_and_fwd_flag_raw))
            WHEN 'Y' THEN TRUE
            WHEN 'N' THEN FALSE
        END                                                              AS store_and_fwd,
        NULLIF(TRIM(pu_location_id_raw), '')::SMALLINT                  AS pu_location_id,
        NULLIF(TRIM(do_location_id_raw), '')::SMALLINT                  AS do_location_id,
        NULLIF(TRIM(payment_type_raw), '')::SMALLINT                    AS payment_type,
        NULLIF(TRIM(fare_amount_raw), '')::NUMERIC(10, 2)               AS fare_amount,
        NULLIF(TRIM(extra_raw), '')::NUMERIC(10, 2)                     AS extra,
        NULLIF(TRIM(mta_tax_raw), '')::NUMERIC(10, 2)                   AS mta_tax,
        NULLIF(TRIM(tip_amount_raw), '')::NUMERIC(10, 2)                AS tip_amount,
        NULLIF(TRIM(tolls_amount_raw), '')::NUMERIC(10, 2)              AS tolls_amount,
        NULLIF(TRIM(improvement_surcharge_raw), '')::NUMERIC(10, 2)     AS improvement_surcharge,
        NULLIF(TRIM(total_amount_raw), '')::NUMERIC(10, 2)              AS total_amount,
        NULLIF(TRIM(congestion_surcharge_raw), '')::NUMERIC(10, 2)      AS congestion_surcharge,
        NULLIF(TRIM(airport_fee_raw), '')::NUMERIC(10, 2)               AS airport_fee
    FROM hw3_taxi_staging
)
INSERT INTO hw3_taxi_trips (
    vendor_id, pickup_at, dropoff_at, passenger_count, trip_distance,
    rate_code_id, store_and_fwd, pu_location_id, do_location_id, payment_type,
    fare_amount, extra, mta_tax, tip_amount, tolls_amount,
    improvement_surcharge, total_amount, congestion_surcharge, airport_fee
)
SELECT
    vendor_id, pickup_at, dropoff_at, passenger_count, trip_distance,
    rate_code_id, store_and_fwd, pu_location_id, do_location_id, payment_type,
    fare_amount, extra, mta_tax, tip_amount, tolls_amount,
    improvement_surcharge, total_amount, congestion_surcharge, airport_fee
FROM parsed
WHERE pickup_at  IS NOT NULL
  AND dropoff_at IS NOT NULL
  AND dropoff_at >= pickup_at;
""")

q("""
SELECT trip_id, vendor_id, pickup_at, dropoff_at, pickup_date, passenger_count,
       trip_distance, rate_code_id, store_and_fwd, payment_type, fare_amount, total_amount
FROM hw3_taxi_trips
ORDER BY trip_id
LIMIT 5
""")

,trip_id,vendor_id,pickup_at,dropoff_at,pickup_date,passenger_count,trip_distance,rate_code_id,store_and_fwd,payment_type,fare_amount,total_amount
0,1,2,2023-12-31 23:56:46-05:00,2024-01-01 00:12:06-05:00,2023-12-31,2,2.38,1,False,1,15.6,21.60
1,2,2,2023-12-31 23:58:35-05:00,2024-01-01 00:13:06-05:00,2023-12-31,6,8.39,1,False,2,33.1,42.35
2,3,2,2024-01-01 00:00:17-05:00,2024-01-01 00:07:46-05:00,2024-01-01,1,2.35,1,False,1,12.1,20.52
3,4,2,2024-01-01 00:01:39-05:00,2024-01-01 01:02:53-05:00,2024-01-01,1,2.59,1,False,2,47.1,52.10
4,5,2,2024-01-01 00:02:35-05:00,2024-01-01 00:10:55-05:00,2024-01-01,1,1.36,1,False,1,10.0,18.00


Перевірка локалізації часу: сирий рядок `pickup_raw` (місцевий час NY) і той самий момент у `TIMESTAMPTZ`,
показаний у NY та в UTC. У січні NY живе за EST (UTC−5), тож різниця має становити рівно 5 годин.

In [10]:
q("""
SELECT
    s.pickup_raw,
    t.pickup_at                          AS pickup_at_session_tz,
    t.pickup_at AT TIME ZONE 'UTC'       AS pickup_at_utc,
    t.pickup_date
FROM hw3_taxi_trips AS t
JOIN hw3_taxi_staging AS s
  ON (s.pickup_raw::TIMESTAMP AT TIME ZONE 'America/New_York') = t.pickup_at
 AND s.dropoff_raw::TIMESTAMP AT TIME ZONE 'America/New_York' = t.dropoff_at
ORDER BY t.trip_id
LIMIT 3
""")

,pickup_raw,pickup_at_session_tz,pickup_at_utc,pickup_date
0,2023-12-31 23:56:46,2023-12-31 23:56:46-05:00,2024-01-01 04:56:46,2023-12-31
1,2023-12-31 23:58:35,2023-12-31 23:58:35-05:00,2024-01-01 04:58:35,2023-12-31
2,2024-01-01 00:00:17,2024-01-01 00:00:17-05:00,2024-01-01 05:00:17,2024-01-01


## Завдання 3. Data-quality audit

### 3.1. Контроль кількості рядків

In [11]:
q("""
SELECT 'staging' AS layer, COUNT(*) AS n_rows FROM hw3_taxi_staging
UNION ALL
SELECT 'clean'   AS layer, COUNT(*) AS n_rows FROM hw3_taxi_trips
""")

,layer,n_rows
0,staging,100000
1,clean,99999


Які саме рядки відфільтровано і чому:

In [12]:
q("""
SELECT
    vendor_id_raw, pickup_raw, dropoff_raw, trip_distance_raw, fare_amount_raw, total_amount_raw
FROM hw3_taxi_staging
WHERE pickup_raw IS NULL
   OR dropoff_raw IS NULL
   OR dropoff_raw::TIMESTAMP < pickup_raw::TIMESTAMP
""")

,vendor_id_raw,pickup_raw,dropoff_raw,trip_distance_raw,fare_amount_raw,total_amount_raw
0,1,2024-01-18 11:00:00,2024-01-18 10:57:07,1.7,19.5,21.0


**Висновок:** У staging завантажено 100 000 рядків, у clean — 99 999. Відфільтровано рівно **1 рядок** за правилом `dropoff_at < pickup_at`: поїздка вендора 1 від 18.01.2024, де висадка (10:57:07) раніше за посадку (11:00:00). Це фізично неможлива подія (ймовірно, збій годинника таксометра), і вона порушила б `CHECK (dropoff_at >= pickup_at)`. Рядків із нерозпізнаним часом (`NULL` після парсингу) немає. Усі інші підозрілі записи свідомо залишено в clean і розглянуто в аудиті нижче.

### 3.2. Перевірка дублікатів

Природного унікального ID поїздки у TLC-даних немає. Як **business key** беремо комбінацію, яка практично
однозначно описує поїздку: `vendor_id` + `pickup_at` + `dropoff_at` + `pu_location_id` + `do_location_id` +
`trip_distance` + `total_amount`. Два різні реальні рейси одного вендора з однаковим часом до секунди, маршрутом,
відстанню та сумою вкрай малоймовірні, тож збіг ключа = потенційний технічний дубль. Це перевірка лише **точних/потенційних**
дублікатів: вона не знайде «майже однакові» записи (наприклад, зсунуті на секунду).

In [13]:
q("""
SELECT
    COUNT(*) AS total_rows,
    COUNT(DISTINCT (vendor_id, pickup_at, dropoff_at, pu_location_id, do_location_id,
                    trip_distance, total_amount)) AS distinct_business_keys,
    COUNT(*) - COUNT(DISTINCT (vendor_id, pickup_at, dropoff_at, pu_location_id, do_location_id,
                               trip_distance, total_amount)) AS duplicate_rows
FROM hw3_taxi_trips
""")

,total_rows,distinct_business_keys,duplicate_rows
0,99999,99999,0


Слабший ключ — без сум (той самий рейс, але з різними фінансовими записами, наприклад оригінал + коригування):

In [14]:
q("""
SELECT
    vendor_id, pickup_at, dropoff_at, pu_location_id, do_location_id,
    COUNT(*)                                  AS n_records,
    STRING_AGG(total_amount::TEXT, ' | ')     AS total_amounts,
    STRING_AGG(payment_type::TEXT, ' | ')     AS payment_types
FROM hw3_taxi_trips
GROUP BY vendor_id, pickup_at, dropoff_at, pu_location_id, do_location_id
HAVING COUNT(*) > 1
ORDER BY n_records DESC, pickup_at
LIMIT 10
""")

,vendor_id,pickup_at,dropoff_at,pu_location_id,do_location_id,n_records,total_amounts,payment_types
0,2,2024-01-01 03:50:10-05:00,2024-01-01 04:04:50-05:00,230,143,2,22.00 | -22.00,4 | 4
1,2,2024-01-03 02:32:50-05:00,2024-01-03 02:43:28-05:00,79,195,2,-26.20 | 26.20,4 | 4
2,2,2024-01-04 22:16:02-05:00,2024-01-04 22:22:30-05:00,162,186,2,-12.90 | 12.90,4 | 4
3,2,2024-01-05 09:53:08-05:00,2024-01-05 09:53:16-05:00,143,143,2,-7.00 | 7.00,4 | 4
4,2,2024-01-06 01:42:36-05:00,2024-01-06 01:48:32-05:00,230,246,2,12.90 | -12.90,4 | 4
5,2,2024-01-06 21:35:02-05:00,2024-01-06 21:38:40-05:00,114,114,2,-10.10 | 10.10,4 | 4
6,2,2024-01-06 22:15:05-05:00,2024-01-06 22:18:26-05:00,43,163,2,10.10 | -10.10,4 | 4
7,2,2024-01-07 16:43:21-05:00,2024-01-07 16:44:11-05:00,48,48,2,-7.00 | 7.00,2 | 2
8,2,2024-01-08 16:39:14-05:00,2024-01-08 16:42:56-05:00,24,238,2,-9.80 | 9.80,3 | 3
9,2,2024-01-12 22:07:26-05:00,2024-01-12 22:07:32-05:00,43,237,2,-74.00 | 74.00,2 | 2


**Висновок:** За повним business key (вендор + обидва моменти + зони + відстань + сума) **точних дублікатів немає** (0 з 99 999). Проте слабший ключ без сум знаходить пари «той самий рейс — два фінансові записи», де суми дзеркальні: `22.00 | -22.00`, `-74.00 | 74.00` тощо, переважно з `payment_type` 4 (Dispute), 3 (No charge) або 2. Це не технічні дублікати, а **оригінальна транзакція + сторнування**: у сумі такий рейс дає нульову виручку. Для аналізу виручки їх можна залишити (вони взаємно гасяться), але для моделей «вартість поїздки» обидва записи треба виключити або згорнути. Оскільки гарантованого природного ключа немає, перевірка виявляє лише точні/потенційні дублікати — «майже однакові» записи вона не ловить.

### 3.3. Null-rate і спеціальні маркери

Спочатку — null-rate у **clean**-таблиці по ключових колонках (`rate_code_id` тут уже містить NULL,
отримані з маркера `99`).

In [15]:
q("""
SELECT
    COUNT(*) AS total_rows,
    ROUND(COUNT(*) FILTER (WHERE passenger_count      IS NULL) * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_passenger_count,
    ROUND(COUNT(*) FILTER (WHERE rate_code_id         IS NULL) * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_rate_code_id,
    ROUND(COUNT(*) FILTER (WHERE store_and_fwd        IS NULL) * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_store_and_fwd,
    ROUND(COUNT(*) FILTER (WHERE congestion_surcharge IS NULL) * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_congestion,
    ROUND(COUNT(*) FILTER (WHERE airport_fee          IS NULL) * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_airport_fee,
    ROUND(COUNT(*) FILTER (WHERE trip_distance        IS NULL) * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_trip_distance,
    ROUND(COUNT(*) FILTER (WHERE fare_amount          IS NULL) * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_fare_amount,
    ROUND(COUNT(*) FILTER (WHERE pickup_at            IS NULL) * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_pickup_at
FROM hw3_taxi_trips
""")

,total_rows,pct_null_passenger_count,pct_null_rate_code_id,pct_null_store_and_fwd,pct_null_congestion,pct_null_airport_fee,pct_null_trip_distance,pct_null_fare_amount,pct_null_pickup_at
0,99999,4.73,5.69,4.73,4.73,4.73,0.0,0.0,0.0


Тепер — **сирі** дані в staging: розрізняємо справжній SQL `NULL`, порожній рядок `''` і доменні маркери,
що означають «невідомо», хоча формально є значенням:
`RatecodeID = 99` (Null/unknown за словником), `payment_type = 0` (коду немає у словнику),
`payment_type = 5` (Unknown), `passenger_count = 0`, зони `264/265` (Unknown / Outside of NYC).

In [16]:
q("""
SELECT
    COUNT(*)                                                                  AS total_rows,
    COUNT(*) FILTER (WHERE passenger_count_raw IS NULL)                        AS passenger_sql_null,
    COUNT(*) FILTER (WHERE TRIM(passenger_count_raw) = '')                     AS passenger_empty_string,
    COUNT(*) FILTER (WHERE passenger_count_raw::NUMERIC = 0)                   AS passenger_zero,
    COUNT(*) FILTER (WHERE rate_code_id_raw IS NULL)                           AS ratecode_sql_null,
    COUNT(*) FILTER (WHERE rate_code_id_raw::NUMERIC = 99)                     AS ratecode_99_unknown,
    COUNT(*) FILTER (WHERE store_and_fwd_flag_raw IS NULL)                     AS store_fwd_sql_null,
    COUNT(*) FILTER (WHERE UPPER(TRIM(store_and_fwd_flag_raw)) NOT IN ('Y', 'N')) AS store_fwd_other_marker,
    COUNT(*) FILTER (WHERE payment_type_raw = '0')                             AS payment_code_0,
    COUNT(*) FILTER (WHERE payment_type_raw = '5')                             AS payment_code_5_unknown,
    COUNT(*) FILTER (WHERE pu_location_id_raw IN ('264', '265')
                        OR do_location_id_raw IN ('264', '265'))               AS unknown_zone_264_265,
    COUNT(*) FILTER (WHERE LOWER(TRIM(vendor_id_raw)) IN ('', '?', 'unknown', 'nan', 'null')) AS vendor_text_markers
FROM hw3_taxi_staging
""")

,total_rows,passenger_sql_null,passenger_empty_string,passenger_zero,ratecode_sql_null,ratecode_99_unknown,store_fwd_sql_null,store_fwd_other_marker,payment_code_0,payment_code_5_unknown,unknown_zone_264_265,vendor_text_markers
0,100000,4727,0,1085,4727,962,4727,0,4727,0,1020,0


Чи збігаються пропуски в різних колонках (один і той самий «бідний» тип запису чи незалежні пропуски)?

In [17]:
q("""
SELECT
    vendor_id,
    payment_type,
    COUNT(*)                                                  AS n_rows,
    COUNT(*) FILTER (WHERE passenger_count IS NULL)           AS null_passenger,
    COUNT(*) FILTER (WHERE store_and_fwd IS NULL)             AS null_store_fwd,
    COUNT(*) FILTER (WHERE congestion_surcharge IS NULL)      AS null_congestion
FROM hw3_taxi_trips
WHERE passenger_count IS NULL
   OR payment_type = 0
GROUP BY vendor_id, payment_type
ORDER BY n_rows DESC
""")

,vendor_id,payment_type,n_rows,null_passenger,null_store_fwd,null_congestion
0,2,0,3098,3098,3098,3098
1,1,0,1623,1623,1623,1623
2,6,0,6,6,6,6


**Висновок:** Пропуски в clean мають **блочну структуру**: `passenger_count`, `store_and_fwd`, `congestion_surcharge`, `airport_fee` — усі рівно 4.73% і приходять з тих самих 4 727 рядків. Останній запит показує, що це повністю рядки з `payment_type = 0` у всіх трьох вендорів: такі записи надходять без метаданих поїздки, тобто це один «бідний» тип записів, а не незалежні пропуски. `rate_code_id` має більше NULL (5.69%), бо до 4 727 справжніх `NULL` додалися **962 рядки з маркером `99`**, який за словником означає «Null/unknown» і нормалізований у SQL `NULL`. Порожніх рядків `''` та текстових маркерів (`?`, `unknown`, `nan`) у сирих даних немає — pandas записав пропуски порожніми полями, а COPY перетворив їх на `NULL`. Окремо виявлено значення, які формально не NULL, але по суті означають «невідомо»: `passenger_count = 0` (1 085 рядків), зони 264/265 (1 020 рядків), код оплати `0`, якого немає в словнику. Ключові поля (`pickup_at`, `trip_distance`, `fare_amount`) заповнені на 100%, але це ще не гарантує якості — нижче видно, що в них є від'ємні й неправдоподібні значення.

## Завдання 4. Розподіли та потенційні аномалії

### 4.1. Розподіл категоріальної колонки: `payment_type`

In [18]:
q("""
SELECT
    payment_type,
    CASE payment_type
        WHEN 0 THEN 'Not in dictionary (0)'
        WHEN 1 THEN 'Credit card'
        WHEN 2 THEN 'Cash'
        WHEN 3 THEN 'No charge'
        WHEN 4 THEN 'Dispute'
        WHEN 5 THEN 'Unknown'
        WHEN 6 THEN 'Voided trip'
    END AS payment_label,
    COUNT(*) AS n_rows,
    ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (), 2) AS pct_rows
FROM hw3_taxi_trips
GROUP BY payment_type
ORDER BY n_rows DESC, payment_type
LIMIT 10
""")

,payment_type,payment_label,n_rows,pct_rows
0,1,Credit card,78131,78.13
1,2,Cash,14877,14.88
2,0,Not in dictionary (0),4727,4.73
3,4,Dispute,1594,1.59
4,3,No charge,670,0.67


**Висновок:** Розподіл дуже незбалансований: **картка — 78.13%**, готівка — 14.88%. Код `0` (4.73%) відсутній у словнику і збігається з блоком рядків без метаданих — його треба трактувати як «невідомий спосіб оплати», а не як окрему бізнес-категорію. Рідкісні категорії Dispute (1.59%) і No charge (0.67%) — це в основному сторнування й безкоштовні рейси; коди 5 (Unknown) і 6 (Voided) у вибірці не трапились зовсім. Для ML це означає: `payment_type` не можна використовувати як ознаку в задачі передбачення чайових без урахування того, що готівкові чайові не фіксуються.

Другий домен-важливий категоріальний розподіл — тариф `rate_code_id` (NULL = невідомий, включно з колишнім маркером 99):

In [19]:
q("""
SELECT
    rate_code_id,
    COUNT(*) AS n_rows,
    ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (), 2) AS pct_rows
FROM hw3_taxi_trips
GROUP BY rate_code_id
ORDER BY n_rows DESC, rate_code_id
LIMIT 10
""")

,rate_code_id,n_rows,pct_rows
0,1.0,89828,89.83
1,NaN,5688,5.69
2,2.0,3351,3.35
3,5.0,657,0.66
4,3.0,244,0.24
5,4.0,231,0.23


**Висновок:** Стандартний тариф домінує (89.83%). Другий за частотою — невідомий тариф (`NULL`, 5.69%), далі JFK flat fare (3.35%). Negotiated, Newark, Nassau/Westchester — менше 1% кожен, а Group ride (6) відсутній. Рідкісні тарифи — валідні, а не помилкові: вони мають зовсім іншу логіку ціни (фіксована ставка JFK, домовленість), тому в моделі вартості їх краще тримати окремою категорією, а не зливати в 'other'.

### 4.2. Базова статистика числових колонок: `trip_distance`, `fare_amount`, `total_amount`, тривалість

In [20]:
q("""
SELECT 'trip_distance' AS metric,
       COUNT(trip_distance) AS n_non_null, MIN(trip_distance) AS min_value, MAX(trip_distance) AS max_value,
       ROUND(AVG(trip_distance), 2) AS avg_value, ROUND(STDDEV_SAMP(trip_distance), 2) AS sd_value,
       PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY trip_distance) AS median_value
FROM hw3_taxi_trips
UNION ALL
SELECT 'fare_amount', COUNT(fare_amount), MIN(fare_amount), MAX(fare_amount),
       ROUND(AVG(fare_amount), 2), ROUND(STDDEV_SAMP(fare_amount), 2),
       PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY fare_amount)
FROM hw3_taxi_trips
UNION ALL
SELECT 'total_amount', COUNT(total_amount), MIN(total_amount), MAX(total_amount),
       ROUND(AVG(total_amount), 2), ROUND(STDDEV_SAMP(total_amount), 2),
       PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY total_amount)
FROM hw3_taxi_trips
UNION ALL
SELECT 'duration_min', COUNT(*),
       ROUND(MIN(EXTRACT(EPOCH FROM dropoff_at - pickup_at) / 60), 2),
       ROUND(MAX(EXTRACT(EPOCH FROM dropoff_at - pickup_at) / 60), 2),
       ROUND(AVG(EXTRACT(EPOCH FROM dropoff_at - pickup_at) / 60), 2),
       ROUND(STDDEV_SAMP(EXTRACT(EPOCH FROM dropoff_at - pickup_at) / 60), 2),
       PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY EXTRACT(EPOCH FROM dropoff_at - pickup_at) / 60)
FROM hw3_taxi_trips
""")

,metric,n_non_null,min_value,max_value,avg_value,sd_value,median_value
0,trip_distance,99999,0.00,99.00,3.21,4.33,1.680000
1,total_amount,99999,-188.71,623.71,26.68,22.67,20.100000
2,fare_amount,99999,-170.00,591.70,18.09,18.01,12.800000
3,duration_min,99999,0.00,1439.10,15.32,29.13,11.616667


**Висновок:** Усі метрики сильно скошені вправо: медіана відстані 1.68 милі проти середнього 3.21, медіанна тривалість 11.6 хв проти середніх 15.3 хв при SD 29 хв. Мінімуми одразу показують проблеми: `fare_amount` і `total_amount` бувають від'ємні (до −170 / −188.71 USD), відстань і тривалість бувають нульові, а максимальна тривалість 1 439 хв (≈24 год) для поїздки таксі неправдоподібна. Максимальна відстань 99 миль і чек 623.71 USD — рідкісні, але можливі (далекі виїзди за межі міста). Через скошеність середнє та SD погано описують «типову» поїздку — для звітів краще медіана і перцентилі.

### 4.3. Outlier-check (доменний)

Розподіли `trip_distance`, `fare_amount` і тривалості сильно скошені вправо (див. середнє vs медіана вище),
тому правило 3σ тут некоректне — воно розраховане на приблизно симетричні дані. Замість нього —
**доменні правила**, кожне з яких має фізичний або бізнес-зміст:

- `fare_amount < 0` або `total_amount < 0` — від'ємний чек;
- `trip_distance > 100` миль — неправдоподібно для поїздки NYC-таксі;
- тривалість `> 5` годин;
- середня швидкість `> 80` mph (при відстані > 1 милі);
- `trip_distance = 0`, але `fare_amount > 50` — велика оплата без руху.

In [21]:
q("""
WITH flagged AS (
    SELECT
        trip_id, vendor_id, pickup_at, dropoff_at, pu_location_id, do_location_id,
        payment_type, trip_distance, fare_amount, total_amount,
        ROUND(EXTRACT(EPOCH FROM dropoff_at - pickup_at) / 60, 1) AS duration_min,
        CASE
            WHEN fare_amount < 0 OR total_amount < 0                        THEN 'negative_amount'
            WHEN trip_distance > 100                                        THEN 'distance_gt_100mi'
            WHEN dropoff_at - pickup_at > INTERVAL '5 hours'                THEN 'duration_gt_5h'
            WHEN trip_distance > 1
             AND trip_distance / NULLIF(EXTRACT(EPOCH FROM dropoff_at - pickup_at) / 3600, 0) > 80
                                                                            THEN 'speed_gt_80mph'
            WHEN trip_distance = 0 AND fare_amount > 50                     THEN 'zero_distance_high_fare'
        END AS anomaly_type
    FROM hw3_taxi_trips
)
SELECT
    anomaly_type,
    COUNT(*)                                   AS n_rows,
    ROUND(COUNT(*) * 100.0 / (SELECT COUNT(*) FROM hw3_taxi_trips), 3) AS pct_of_clean,
    MIN(trip_distance)                         AS min_distance,
    MAX(trip_distance)                         AS max_distance,
    MIN(total_amount)                          AS min_total,
    MAX(total_amount)                          AS max_total,
    MAX(duration_min)                          AS max_duration_min
FROM flagged
WHERE anomaly_type IS NOT NULL
GROUP BY anomaly_type
ORDER BY n_rows DESC
""")

,anomaly_type,n_rows,pct_of_clean,min_distance,max_distance,min_total,max_total,max_duration_min
0,negative_amount,1308,1.308,0.00,31.26,-188.71,3.88,86.6
1,zero_distance_high_fare,293,0.293,0.00,0.00,54.00,578.96,257.9
2,duration_gt_5h,41,0.041,0.43,28.51,9.80,189.28,1439.1
3,speed_gt_80mph,23,0.023,1.10,19.32,0.00,128.40,4.1


Приклади найекстремальніших рядків кожного типу:

In [22]:
q("""
WITH flagged AS (
    SELECT
        trip_id, vendor_id, pickup_at, pu_location_id, do_location_id,
        payment_type, trip_distance, fare_amount, total_amount,
        ROUND(EXTRACT(EPOCH FROM dropoff_at - pickup_at) / 60, 1) AS duration_min,
        CASE
            WHEN fare_amount < 0 OR total_amount < 0                        THEN 'negative_amount'
            WHEN trip_distance > 100                                        THEN 'distance_gt_100mi'
            WHEN dropoff_at - pickup_at > INTERVAL '5 hours'                THEN 'duration_gt_5h'
            WHEN trip_distance > 1
             AND trip_distance / NULLIF(EXTRACT(EPOCH FROM dropoff_at - pickup_at) / 3600, 0) > 80
                                                                            THEN 'speed_gt_80mph'
            WHEN trip_distance = 0 AND fare_amount > 50                     THEN 'zero_distance_high_fare'
        END AS anomaly_type
    FROM hw3_taxi_trips
)
SELECT *
FROM flagged
WHERE anomaly_type IS NOT NULL
  AND anomaly_type <> 'negative_amount'
ORDER BY anomaly_type, trip_distance DESC, duration_min DESC
LIMIT 20
""")

,trip_id,vendor_id,pickup_at,pu_location_id,do_location_id,payment_type,trip_distance,fare_amount,total_amount,duration_min,anomaly_type
0,10489,2,2024-01-04 18:51:46-05:00,132,265,1,28.51,138.1,189.28,308.2,duration_gt_5h
1,25016,2,2024-01-09 15:40:42-05:00,132,158,1,27.47,70.0,132.30,499.3,duration_gt_5h
2,47857,2,2024-01-16 16:42:51-05:00,132,186,1,17.53,70.0,96.46,437.2,duration_gt_5h
3,84052,2,2024-01-27 04:21:38-05:00,48,265,1,15.87,59.7,67.94,1434.1,duration_gt_5h
4,11749,2,2024-01-05 06:22:11-05:00,132,132,2,10.37,41.5,44.75,1057.8,duration_gt_5h
5,88218,2,2024-01-28 10:43:35-05:00,161,138,1,10.29,44.3,63.96,796.4,duration_gt_5h
6,5861,2,2024-01-03 10:50:34-05:00,138,161,1,10.28,47.8,75.31,1353.0,duration_gt_5h
7,83309,2,2024-01-26 23:04:09-05:00,142,33,1,6.99,35.2,48.24,1435.3,duration_gt_5h
8,83579,2,2024-01-27 00:24:34-05:00,132,132,2,5.78,24.7,28.95,973.1,duration_gt_5h
9,17063,2,2024-01-06 17:03:33-05:00,116,163,1,5.71,28.2,35.42,1434.0,duration_gt_5h


Від'ємні суми — з якими `payment_type` вони пов'язані?

In [23]:
q("""
SELECT
    payment_type,
    COUNT(*)                 AS n_negative,
    ROUND(AVG(total_amount), 2) AS avg_total,
    MIN(total_amount)        AS min_total
FROM hw3_taxi_trips
WHERE fare_amount < 0 OR total_amount < 0
GROUP BY payment_type
ORDER BY n_negative DESC
""")

,payment_type,n_negative,avg_total,min_total
0,4,766,-25.19,-188.71
1,2,273,-27.27,-129.80
2,3,205,-20.96,-93.69
3,0,63,2.49,1.00
4,1,1,-14.30,-14.30


**Висновок і запропоновані дії:** Знайдено 4 типи аномалій (разом ≈1.7% рядків), і дії для них різні:

- **`negative_amount` — 1 308 рядків (1.31%).** Здебільшого payment_type 4 (Dispute, 766), 2 (Cash, 273) і 3 (No charge, 205) — це сторнування/повернення, частина яких має дзеркальну пару (див. 3.2). **Дія:** залишити в даних для фінансової звірки, додати прапорець `is_reversal` і **виключити з training data** для моделей вартості. Окремо 63 рядки з кодом `0`, де `fare_amount` від'ємний, а `total_amount` додатний — це проблема формування чека, їх варто **передати на перевірку джерела**.
- **`zero_distance_high_fare` — 293 рядки.** Оплата понад 50 USD при нульовій відстані: частина — Negotiated fare, де відстань не рахується, частина — збій одометра. **Дія:** додати прапорець, не використовувати `trip_distance` для цих рядків (можна відновити відстань за парою зон).
- **`duration_gt_5h` — 41 рядок.** Більшість має тривалість ≈1 400–1 439 хв при звичайній відстані й сумі (наприклад, 1.89 милі за 19.68 USD за «23.8 год»). Майже всі — вендор 2. Це типовий збій: таксометр не закрив поїздку до наступної доби. **Дія:** виправити неможливо, тому прапорець + виключити з training data для моделі тривалості; передати вендору на перевірку.
- **`speed_gt_80mph` — 23 рядки.** Відстань у кілька миль за 1–4 хв — помилка часу або GPS. **Дія:** прапорець і виключення з моделей швидкості/тривалості.

Великі суми та відстані самі по собі (99 миль, 623 USD) **не видаляються** — це рідкісні, але валідні поїздки за межі міста.

## Завдання 5. Десять DQL/EDA-запитів

Покриття обов'язкових конструкцій:

| Конструкція | Запити |
|---|---|
| явний перелік колонок у `SELECT` | усі (Q1–Q10) |
| `WHERE` з `AND` / `OR` / `NOT` і дужками | Q4, Q8 |
| `IN` / `BETWEEN` | Q5 (`IN`), Q8 (`BETWEEN`), Q9 (`IN`) |
| `LIKE` / `ILIKE` | Q5 (`ILIKE`), Q9 (`LIKE`) |
| `IS NULL` / `IS NOT NULL` | Q3, Q6, Q10 |
| `COALESCE` / `NULLIF` | Q3 (`COALESCE`), Q7 (`COALESCE`), Q10 (`NULLIF`) |
| `IS DISTINCT FROM` / `IS NOT DISTINCT FROM` | Q7, Q9 |
| `DISTINCT` | Q6 |
| `ORDER BY` + `LIMIT` | Q1, Q4, Q7, Q8, Q9 |
| `GROUP BY` | Q1, Q2, Q3, Q5, Q6, Q9, Q10 |

### Q1
**Business-question:** Які способи оплати найпоширеніші і скільки в середньому платить пасажир за кожним із них?

In [24]:
q("""
SELECT
    payment_type,
    COUNT(*)                      AS n_trips,
    ROUND(AVG(total_amount), 2)   AS avg_total,
    ROUND(AVG(tip_amount), 2)     AS avg_tip
FROM hw3_taxi_trips
GROUP BY payment_type
ORDER BY n_trips DESC
LIMIT 5
""")

,payment_type,n_trips,avg_total,avg_tip
0,1,78131,28.17,4.15
1,2,14877,22.81,0.00
2,0,4727,25.64,1.54
3,4,1594,0.59,0.05
4,3,670,7.24,0.01


**Interpretation:** Картою оплачено 78 131 поїздку із середнім чеком 28.17 USD і чайовими 4.15 USD, готівкою — 14 877 поїздок із чеком 22.81 USD і **чайовими рівно 0.00**. Нуль у готівкових чайових — не поведінка пасажирів, а обмеження збору даних: готівкові чайові таксометр не фіксує. Середній чек у Dispute (0.59 USD) і No charge (7.24 USD) близький до нуля, бо ці категорії містять сторнування з від'ємними сумами. Отже, модель чайових можна будувати лише на `payment_type = 1`.

### Q2
**Business-question:** У які години доби (місцевий час Нью-Йорка) відбувається найбільше поїздок?

In [25]:
q("""
SELECT
    EXTRACT(HOUR FROM pickup_at AT TIME ZONE 'America/New_York')::INT AS pickup_hour,
    COUNT(*)                                                         AS n_trips,
    ROUND(AVG(trip_distance), 2)                                     AS avg_distance
FROM hw3_taxi_trips
GROUP BY pickup_hour
ORDER BY pickup_hour
""")

,pickup_hour,n_trips,avg_distance
0,0,2670,3.67
1,1,1810,3.10
2,2,1233,2.86
3,3,817,3.38
4,4,559,4.35
5,5,638,6.37
6,6,1405,4.52
7,7,2774,3.42
8,8,3817,2.94
9,9,4379,3.00


**Interpretation:** Пік попиту — **17:00–18:00** (6 946 і 7 082 поїздки), загалом період 14:00–19:00 дає понад 6 000 поїздок щогодини. Мінімум — о 4:00 (559 поїздок), тобто в 12.7 раза менше, ніж у пік. Цікаво, що о 5:00 середня відстань максимальна (6.37 милі проти ~3 у день): ранні поїздки непропорційно часто їдуть в аеропорти. Година доби — сильна ознака для моделей попиту й тривалості; важливо, що вона обчислена в місцевому часі NY, а не в UTC.

### Q3
**Business-question:** Яка частка поїздок має невідому кількість пасажирів і як розподілені поїздки за `passenger_count`?

In [26]:
q("""
SELECT
    COALESCE(passenger_count::TEXT, 'unknown (NULL)')                 AS passenger_group,
    COUNT(*)                                                          AS n_trips,
    ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (), 2)                AS pct_trips,
    ROUND(AVG(total_amount), 2)                                       AS avg_total
FROM hw3_taxi_trips
GROUP BY passenger_count
ORDER BY passenger_count IS NULL, passenger_count
""")

,passenger_group,n_trips,pct_trips,avg_total
0,0,1085,1.09,24.21
1,1,73738,73.74,26.10
2,2,13809,13.81,29.46
3,3,3071,3.07,28.93
4,4,1722,1.72,30.76
5,5,1140,1.14,25.50
6,6,705,0.71,25.14
7,8,2,0.00,107.70
8,unknown (NULL),4727,4.73,25.64


**Interpretation:** Кількість пасажирів невідома в **4.73%** поїздок (4 727 рядків — той самий блок `payment_type = 0`). Ще 1.09% мають `passenger_count = 0`, що фізично неможливо для оплаченої поїздки — це неповний ввід водія, і його варто трактувати як пропуск. Переважна більшість — поїздки з 1 пасажиром (73.74%), з 2 — 13.81%; значення 7–9 практично відсутні (2 поїздки з 8 пасажирами — ймовірно, групові рейси, середній чек 107.70 USD). `COALESCE` дозволив явно показати NULL як окрему групу, а не втратити його в агрегації.

### Q4
**Business-question:** Які поїздки мають від'ємні суми або нульову відстань, але при цьому **не** є офіційно анульованими/спірними (payment_type 3, 4, 6)?

In [27]:
q("""
SELECT
    trip_id,
    pickup_at,
    payment_type,
    rate_code_id,
    trip_distance,
    fare_amount,
    total_amount,
    pu_location_id,
    do_location_id
FROM hw3_taxi_trips
WHERE (fare_amount < 0 OR total_amount < 0 OR trip_distance = 0)
  AND NOT (payment_type IN (3, 4, 6))
ORDER BY total_amount ASC
LIMIT 10
""")

,trip_id,pickup_at,payment_type,rate_code_id,trip_distance,fare_amount,total_amount,pu_location_id,do_location_id
0,6550,2024-01-03 14:57:51-05:00,2,3,20.52,-99.30,-129.80,230,1
1,5401,2024-01-03 07:13:26-05:00,2,1,24.67,-96.10,-100.10,137,39
2,4520,2024-01-02 18:12:04-05:00,2,2,18.60,-70.00,-87.69,132,230
3,10629,2024-01-04 19:16:44-05:00,2,2,16.68,-70.00,-87.69,132,164
4,10571,2024-01-04 19:07:22-05:00,2,2,18.44,-70.00,-87.69,132,163
5,92192,2024-01-29 17:25:37-05:00,2,4,15.27,-70.90,-84.15,138,265
6,61215,2024-01-20 13:44:21-05:00,2,5,3.72,-82.69,-83.69,216,132
7,91607,2024-01-29 14:22:04-05:00,2,2,17.57,-70.00,-83.44,68,132
8,60935,2024-01-20 12:25:54-05:00,2,2,4.53,-70.00,-82.94,260,137
9,62183,2024-01-20 18:07:57-05:00,2,2,22.26,-70.00,-82.69,132,48


Скільки таких рядків загалом, окремо за кожним правилом:

In [28]:
q("""
SELECT
    COUNT(*) FILTER (WHERE (fare_amount < 0 OR total_amount < 0) AND NOT (payment_type IN (3, 4, 6))) AS negative_not_disputed,
    COUNT(*) FILTER (WHERE trip_distance = 0 AND NOT (payment_type IN (3, 4, 6)))                   AS zero_distance_not_disputed,
    COUNT(*) FILTER (WHERE trip_distance = 0 AND rate_code_id = 5)                                  AS zero_distance_negotiated
FROM hw3_taxi_trips
""")

,negative_not_disputed,zero_distance_not_disputed,zero_distance_negotiated
0,337,1703,346


**Interpretation:** Серед поїздок, які **не** позначені як Dispute / No charge / Voided, є 337 рядків з від'ємними сумами і 1 703 — з нульовою відстанню. Найбільші від'ємні чеки (до −129.80 USD) мають `payment_type = 2` (Cash) і часто тариф 2 (JFK, fare −70.00) — це сторнування, які не були правильно позначені кодом оплати. Частину нульових відстаней пояснює тариф Negotiated (5), де відстань не має значення для ціни: таких поїздок 346. Отже, `payment_type` не можна використовувати як надійний індикатор сторнування — потрібна окрема ознака `total_amount < 0`.

### Q5
**Business-question:** Наскільки поїздки з аеропортів (JFK, LaGuardia, Newark) відрізняються від решти за відстанню, вартістю та чайовими?

In [29]:
q("""
SELECT
    CASE
        WHEN pu_location_id IN (SELECT location_id FROM hw3_taxi_zones WHERE zone ILIKE '%airport%')
        THEN 'airport pickup'
        ELSE 'other pickup'
    END                                  AS pickup_kind,
    COUNT(*)                             AS n_trips,
    ROUND(AVG(trip_distance), 2)         AS avg_distance,
    ROUND(AVG(total_amount), 2)          AS avg_total,
    ROUND(AVG(tip_amount), 2)            AS avg_tip,
    ROUND(AVG(airport_fee), 2)           AS avg_airport_fee
FROM hw3_taxi_trips
GROUP BY pickup_kind
ORDER BY n_trips DESC
""")

,pickup_kind,n_trips,avg_distance,avg_total,avg_tip,avg_airport_fee
0,other pickup,92231,2.37,22.86,2.86,0.01
1,airport pickup,7768,13.15,71.93,8.83,1.62


Які саме зони потрапили під фільтр `ILIKE '%airport%'`:

In [30]:
q("""
SELECT location_id, borough, zone, service_zone
FROM hw3_taxi_zones
WHERE zone ILIKE '%airport%'
ORDER BY location_id
""")

,location_id,borough,zone,service_zone
0,1,EWR,Newark Airport,EWR
1,132,Queens,JFK Airport,Airports
2,138,Queens,LaGuardia Airport,Airports


**Interpretation:** Поїздки з аеропортів — 7 768 з 99 999 (≈7.8%), але вони **в 5.5 раза довші** (13.15 проти 2.37 милі) і **втричі дорожчі** (71.93 проти 22.86 USD), з утричі більшими чайовими (8.83 проти 2.86 USD). `ILIKE '%airport%'` коректно знайшов три зони: Newark, JFK і LaGuardia. Середній `airport_fee` 1.62 (а не 1.75), бо за правилами TLC цей збір стягується лише при посадці в LaGuardia та JFK, а не в Newark. Для моделей вартості ознака «аеропортна поїздка» критично важлива.

### Q6
**Business-question:** Які унікальні комбінації вендора і тарифу трапляються в даних і чи пов'язані невідомі тарифи з конкретним вендором?

In [31]:
q("""
SELECT DISTINCT
    vendor_id,
    rate_code_id,
    rate_code_id IS NULL AS is_unknown_rate
FROM hw3_taxi_trips
ORDER BY vendor_id, rate_code_id
""")

,vendor_id,rate_code_id,is_unknown_rate
0,1,1.0,False
1,1,2.0,False
2,1,3.0,False
3,1,4.0,False
4,1,5.0,False
5,1,NaN,True
6,2,1.0,False
7,2,2.0,False
8,2,3.0,False
9,2,4.0,False


In [32]:
q("""
SELECT
    vendor_id,
    COUNT(*)                                            AS n_trips,
    COUNT(*) FILTER (WHERE rate_code_id IS NULL)        AS unknown_rate,
    COUNT(*) FILTER (WHERE rate_code_id IS NOT NULL)    AS known_rate,
    COUNT(DISTINCT rate_code_id)                        AS n_distinct_known_rates
FROM hw3_taxi_trips
GROUP BY vendor_id
ORDER BY vendor_id
""")

,vendor_id,n_trips,unknown_rate,known_rate,n_distinct_known_rates
0,1,24784,2584,22200,5
1,2,75209,3098,72111,5
2,6,6,6,0,0


**Interpretation:** Вендори 1 і 2 використовують однаковий набір тарифів 1–5 і обидва мають невідомий тариф (2 584 і 3 098 рядків відповідно), тобто пропуск тарифу не прив'язаний до одного вендора. Вендор 6 (Myle) представлений лише 6 поїздками і **всі** вони без тарифу — його дані в цьому місяці неповні і мізерні за обсягом. `DISTINCT` одразу показує, що тариф 6 (Group ride) не зустрічається в жодного вендора.

### Q7
**Business-question:** Чи дорівнює `total_amount` сумі своїх складових, і які поїздки мають розбіжність?

Складові: `fare_amount + extra + mta_tax + tip_amount + tolls_amount + improvement_surcharge + congestion_surcharge + airport_fee`.
Пропущені надбавки через `COALESCE` трактуємо як 0; порівняння через `IS DISTINCT FROM`, щоб NULL не «ховав» розбіжності.

In [33]:
q("""
SELECT
    trip_id,
    vendor_id,
    payment_type,
    total_amount,
    fare_amount + COALESCE(extra, 0) + COALESCE(mta_tax, 0) + COALESCE(tip_amount, 0)
        + COALESCE(tolls_amount, 0) + COALESCE(improvement_surcharge, 0)
        + COALESCE(congestion_surcharge, 0) + COALESCE(airport_fee, 0)      AS components_sum,
    total_amount - (fare_amount + COALESCE(extra, 0) + COALESCE(mta_tax, 0) + COALESCE(tip_amount, 0)
        + COALESCE(tolls_amount, 0) + COALESCE(improvement_surcharge, 0)
        + COALESCE(congestion_surcharge, 0) + COALESCE(airport_fee, 0))     AS diff
FROM hw3_taxi_trips
WHERE total_amount IS DISTINCT FROM
      fare_amount + COALESCE(extra, 0) + COALESCE(mta_tax, 0) + COALESCE(tip_amount, 0)
    + COALESCE(tolls_amount, 0) + COALESCE(improvement_surcharge, 0)
    + COALESCE(congestion_surcharge, 0) + COALESCE(airport_fee, 0)
ORDER BY ABS(total_amount - (fare_amount + COALESCE(extra, 0) + COALESCE(mta_tax, 0) + COALESCE(tip_amount, 0)
        + COALESCE(tolls_amount, 0) + COALESCE(improvement_surcharge, 0)
        + COALESCE(congestion_surcharge, 0) + COALESCE(airport_fee, 0))) DESC
LIMIT 10
""")

,trip_id,vendor_id,payment_type,total_amount,components_sum,diff
0,98475,1,0,41.10,36.60,4.5
1,19448,1,0,53.75,49.25,4.5
2,83112,1,0,23.85,19.35,4.5
3,87838,1,0,27.30,22.80,4.5
4,83269,1,0,20.40,15.90,4.5
5,74448,1,0,87.10,82.60,4.5
6,57218,1,0,54.90,50.40,4.5
7,72835,1,0,21.80,17.30,4.5
8,17691,1,0,33.20,28.70,4.5
9,53060,1,0,26.15,21.65,4.5


Розбивка розбіжностей за вендором і розміром різниці — щоб зрозуміти, чи це хаотичні помилки, чи систематичне правило:

In [34]:
q("""
SELECT
    vendor_id,
    total_amount - (fare_amount + COALESCE(extra, 0) + COALESCE(mta_tax, 0) + COALESCE(tip_amount, 0)
        + COALESCE(tolls_amount, 0) + COALESCE(improvement_surcharge, 0)
        + COALESCE(congestion_surcharge, 0) + COALESCE(airport_fee, 0))     AS diff,
    COUNT(*)                                                                AS n_rows,
    COUNT(*) FILTER (WHERE congestion_surcharge IS NULL)                    AS n_null_congestion
FROM hw3_taxi_trips
WHERE total_amount IS DISTINCT FROM
      fare_amount + COALESCE(extra, 0) + COALESCE(mta_tax, 0) + COALESCE(tip_amount, 0)
    + COALESCE(tolls_amount, 0) + COALESCE(improvement_surcharge, 0)
    + COALESCE(congestion_surcharge, 0) + COALESCE(airport_fee, 0)
GROUP BY vendor_id, diff
ORDER BY n_rows DESC
LIMIT 8
""")

,vendor_id,diff,n_rows,n_null_congestion
0,1,-2.50,20141,0
1,2,2.50,2912,2766
2,1,2.50,1492,1492
3,1,-4.25,731,0
4,1,-1.75,641,0
5,1,4.50,18,18
6,2,-2.50,4,0
7,2,-1.75,2,0


In [35]:
q("""
SELECT
    COUNT(*) AS total_rows,
    COUNT(*) FILTER (
        WHERE total_amount IS DISTINCT FROM
              fare_amount + COALESCE(extra, 0) + COALESCE(mta_tax, 0) + COALESCE(tip_amount, 0)
            + COALESCE(tolls_amount, 0) + COALESCE(improvement_surcharge, 0)
            + COALESCE(congestion_surcharge, 0) + COALESCE(airport_fee, 0)
    ) AS mismatched_rows
FROM hw3_taxi_trips
""")

,total_rows,mismatched_rows
0,99999,25941


**Interpretation:** Розбіжність між `total_amount` і сумою складових має **25 941 рядок (≈26%)** — але вона повністю систематична, а не хаотична. У вендора 1 різниця дорівнює рівно −2.50 / −4.25 / −1.75 USD: це точно `congestion_surcharge`, `congestion + airport_fee` і `airport_fee`, тобто вендор 1 **вже включає ці збори в `extra`**, і наша формула рахує їх двічі. Різниця +2.50 (≈4 400 рядків) — це здебільшого записи з `payment_type = 0`, де `congestion_surcharge` = NULL, хоча в `total_amount` збір враховано; лише ~150 рядків вендора 2 мають цю різницю при заповнених надбавках. Висновок: семантика колонки `extra` залежить від вендора, і будь-які ознаки на основі складових чека треба будувати з урахуванням `vendor_id`.

### Q8
**Business-question:** Які найдовші за часом «короткі» поїздки (0.5–2 милі) у робочі дні 8–20 січня з оплатою карткою або готівкою — тобто де пасажир найдовше стояв у заторі?

In [36]:
q("""
SELECT
    trip_id,
    pickup_at,
    pickup_date,
    trip_distance,
    ROUND(EXTRACT(EPOCH FROM dropoff_at - pickup_at) / 60, 1) AS duration_min,
    fare_amount,
    pu_location_id,
    do_location_id
FROM hw3_taxi_trips
WHERE trip_distance BETWEEN 0.5 AND 2
  AND pickup_date BETWEEN DATE '2024-01-08' AND DATE '2024-01-20'
  AND NOT (EXTRACT(ISODOW FROM pickup_date) IN (6, 7))
  AND (payment_type = 1 OR payment_type = 2)
  AND dropoff_at - pickup_at < INTERVAL '3 hours'
ORDER BY dropoff_at - pickup_at DESC
LIMIT 10
""")

,trip_id,pickup_at,pickup_date,trip_distance,duration_min,fare_amount,pu_location_id,do_location_id
0,46221,2024-01-16 07:15:37-05:00,2024-01-16,1.90,88.0,43.5,149,233
1,46665,2024-01-16 10:17:18-05:00,2024-01-16,1.52,61.9,46.4,170,163
2,46870,2024-01-16 11:24:20-05:00,2024-01-16,1.40,54.9,40.1,230,233
3,46856,2024-01-16 11:19:48-05:00,2024-01-16,1.70,53.7,40.8,236,162
4,26639,2024-01-10 06:36:15-05:00,2024-01-10,1.60,52.9,17.5,17,49
5,30795,2024-01-11 12:13:22-05:00,2024-01-11,1.60,52.2,32.5,167,161
6,46748,2024-01-16 10:50:12-05:00,2024-01-16,1.80,47.8,28.2,264,264
7,46876,2024-01-16 11:26:44-05:00,2024-01-16,1.69,47.6,36.6,237,161
8,53977,2024-01-18 12:07:42-05:00,2024-01-18,1.70,47.3,11.4,238,166
9,46784,2024-01-16 10:56:49-05:00,2024-01-16,1.01,46.7,35.9,162,170


**Interpretation:** Серед коротких (0.5–2 милі) поїздок у будні найдовше стояли в заторах **16 січня**: 7 з 10 найповільніших рейсів того дня, до 88 хв на 1.9 милі (≈1.3 mph) із тарифом 43.50 USD. Для порівняння, медіанна тривалість поїздки загалом — 11.6 хв. Така концентрація в один день вказує на загальноміську подію (ймовірно, погодна: 16.01.2024 у NYC випав перший за майже два роки помітний сніг), а не на помилки окремих таксометрів. Для моделі тривалості погода і дата — важливі зовнішні ознаки, яких у датасеті немає.

### Q9
**Business-question:** Які маршрути всередині Манхеттена найпопулярніші серед поїздок, що закінчуються в **іншій** зоні, ніж почалися?

`LIKE 'Manhattan'` фільтрує довідник за районом, `IS DISTINCT FROM` відсікає поїздки в межах однієї зони.

In [37]:
q("""
SELECT
    t.pu_location_id,
    pz.zone                         AS pickup_zone,
    t.do_location_id,
    dz.zone                         AS dropoff_zone,
    COUNT(*)                        AS n_trips,
    ROUND(AVG(t.total_amount), 2)   AS avg_total
FROM hw3_taxi_trips AS t
JOIN hw3_taxi_zones AS pz ON pz.location_id = t.pu_location_id
JOIN hw3_taxi_zones AS dz ON dz.location_id = t.do_location_id
WHERE t.pu_location_id IN (SELECT location_id FROM hw3_taxi_zones WHERE borough LIKE 'Manhattan')
  AND t.do_location_id IN (SELECT location_id FROM hw3_taxi_zones WHERE borough LIKE 'Manhattan')
  AND t.pu_location_id IS DISTINCT FROM t.do_location_id
GROUP BY t.pu_location_id, pz.zone, t.do_location_id, dz.zone
ORDER BY n_trips DESC, t.pu_location_id
LIMIT 10
""")

,pu_location_id,pickup_zone,do_location_id,dropoff_zone,n_trips,avg_total
0,237,Upper East Side South,236,Upper East Side North,703,15.86
1,236,Upper East Side North,237,Upper East Side South,639,15.75
2,161,Midtown Center,237,Upper East Side South,325,16.54
3,239,Upper West Side South,238,Upper West Side North,308,13.68
4,237,Upper East Side South,161,Midtown Center,294,16.04
5,239,Upper West Side South,142,Lincoln Square East,292,14.31
6,142,Lincoln Square East,239,Upper West Side South,282,15.13
7,161,Midtown Center,236,Upper East Side North,279,21.23
8,141,Lenox Hill West,236,Upper East Side North,274,15.64
9,236,Upper East Side North,161,Midtown Center,265,20.56


In [38]:
q("""
SELECT
    COUNT(*) FILTER (WHERE pu_location_id IS NOT DISTINCT FROM do_location_id) AS same_zone_trips,
    ROUND(COUNT(*) FILTER (WHERE pu_location_id IS NOT DISTINCT FROM do_location_id) * 100.0 / COUNT(*), 2)
        AS pct_same_zone
FROM hw3_taxi_trips
""")

,same_zone_trips,pct_same_zone
0,5470,5.47


**Interpretation:** Найпопулярніший маршрут усередині Манхеттена — між **Upper East Side South і North** в обох напрямках (703 і 639 поїздок), із середнім чеком ~15.8 USD. Топ-10 складається з коротких поїздок між сусідніми житловими районами (Upper East/West Side) і Midtown Center. Поїздки в межах однієї зони (`IS NOT DISTINCT FROM`) становлять 5.47% — їх виключено, бо вони не описують переміщення між районами. Маршрути, що ведуть у Midtown, дорожчі (20–21 USD) через затори в центрі.

### Q10
**Business-question:** Як відсоток чайових (для поїздок, оплачених карткою) залежить від дня тижня?

`NULLIF(fare_amount, 0)` захищає від ділення на нуль; готівкові чайові в даних не фіксуються, тому беремо лише `payment_type = 1`.

In [39]:
q("""
SELECT
    EXTRACT(ISODOW FROM pickup_date)::INT                          AS iso_dow,
    TO_CHAR(pickup_date, 'Dy')                                     AS dow,
    COUNT(*)                                                       AS n_card_trips,
    ROUND(AVG(tip_amount / NULLIF(fare_amount, 0)) * 100, 2)       AS avg_tip_pct,
    ROUND(COUNT(*) FILTER (WHERE tip_amount = 0) * 100.0 / COUNT(*), 2) AS pct_zero_tip
FROM hw3_taxi_trips
WHERE payment_type = 1
  AND fare_amount > 0
  AND tip_amount IS NOT NULL
GROUP BY iso_dow, dow
ORDER BY iso_dow
""")

,iso_dow,dow,n_card_trips,avg_tip_pct,pct_zero_tip
0,1,Mon,10373,25.63,5.20
1,2,Tue,12274,25.70,4.95
2,3,Wed,13038,25.41,4.93
3,4,Thu,11709,25.85,4.54
4,5,Fri,10729,25.77,5.28
5,6,Sat,11106,26.87,4.75
6,7,Sun,8896,25.34,4.81


**Interpretation:** Середні чайові для оплати карткою стабільні протягом тижня: **25.3–26.9%** від тарифу, найвищі — у суботу (26.87%), найнижчі — у неділю (25.34%). Частка поїздок без чайових також стабільна — 4.5–5.3%. Різниця між днями становить близько 1.5 процентного пункта, тож день тижня — слабка ознака для передбачення чайових. `NULLIF(fare_amount, 0)` захистив від ділення на нуль, хоча додатковий фільтр `fare_amount > 0` уже виключає нульові та від'ємні тарифи.

## Завдання 6. Reflection

**Проблеми якості.** Пропуски тут не випадкові: це блок записів з `payment_type = 0`, у якому одночасно немає пасажирів, тарифу, прапорця та надбавок. Окрім SQL `NULL`, є маркери з тим самим змістом: `RatecodeID = 99`, `passenger_count = 0`, зони 264/265 і код оплати 0, якого немає в словнику. Точних дублікатів немає, але є пари «транзакція + сторнування». Трапляються й неможливі значення: від'ємні суми, поїздки тривалістю 24 години, швидкість понад 80 mph. Найважливіше — `extra` означає різне у різних вендорів, тому `total_amount` не завжди дорівнює сумі своїх складових.

**Ознаки для ML.** Як ознаки варто взяти години й день тижня (у місцевому часі), зони посадки й висадки, `rate_code_id`, `vendor_id`, прапорець аеропорту та `passenger_count` (з NULL як окремою категорією). Не варто використовувати `trip_id`, `store_and_fwd` (технічний атрибут передачі даних) і тип оплати для моделей, що не стосуються чайових.

**Leakage.** Якщо прогнозувати вартість або тривалість у момент посадки, то `dropoff_at`, `trip_distance` (за таксометром), `tip_amount`, `tolls_amount`, `total_amount` і `payment_type` стають відомі лише після завершення поїздки, тож як ознаки їх брати не можна.

**Пропозиції data engineer.** Уніфікувати семантику `extra` між вендорами. Позначати сторнування окремим полем з посиланням на оригінальну поїздку. Замінити маркер 99 і код 0 на справжні NULL або задокументовані коди. Перевіряти `dropoff >= pickup` і розумну тривалість ще до публікації. Додати стабільний ідентифікатор поїздки.

**ML-задача.** Найприродніше тут **regression** (вартість або тривалість поїздки до її початку) та **forecasting** попиту за зонами й годинами. Як допоміжна задача підходить classification (чи дасть пасажир чайові, лише для оплати карткою).